### PR TITLE
Make it a failure when lowering a fusion with issue #369.

### PR DESCRIPTION
**Pull request by naoyam**
The issue of #369 involves allocation and indexing of shared-memory backed tensors when computeAt axes are thread-parallelized. This PR does not fix the issue yet, but it attempts to detect such tensors and throw an exception when detected in order to avoid generating invalid code.

It also adds two test cases that exhibit this issue. Instead of silently generating invalid code, they should throw an exception when lowered to KIR.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 1]
```diff
@@ -24,6 +24,25 @@ LoopNestGenerator::LoopNestGenerator(
   generate(exprs);
 }
 
+namespace {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp:27]: nit: empty lines after namespace { and before the closing }

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 2]
```diff
@@ -24,6 +24,25 @@ LoopNestGenerator::LoopNestGenerator(
...
+// Currently, allocation of smem tennsors and indexing is
+// broken when computeAt axes are thread-parallelized. This check
+// throws an exception if such tensors are detected.
+// TODO: Fix the allocation and indexing of such tensors.
+void failIfUnsupported(TensorView* tv) {
+  for (size_t i = 0; i < tv->getThisComputeAtAxis(); i++) {
+    IterDomain* compute_at_dim = tv->getComputeAtAxis(i).first;
+    const auto memory_type = tv->getMemoryType();
+    if (memory_type == MemoryType::Shared && compute_at_dim->isThreadDim()) {
+      std::stringstream ss;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp:38]: why not build the error msg directly with `TORCH_INTERNAL_ASSERT` ?

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:38]: Printing `TensorView*` with `TORCH_INTERNAL_ASSERT` didn't work for me (just its address was printed). Using `std::stringstream` was a workaround.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:27]: Thanks. Added.

File: test/cpp/jit/test_gpu.cpp [hunk 3]
```diff
@@ -8286,6 +8286,138 @@ TEST(NVFuserTest, FusionNonUniqueBroadcastSize_CUDA) {
...
+    tv->axis(-1)->parallelize(ParallelType::TIDy);
+  }
+
+  // fusion.printMath();
+  // Lowering should throw an error due to tv4 being allocated on
+  // shared memory.
+  ASSERT_ANY_THROW(fusion.printKernel());
+  // Enable the rest of the test
+#if 0
+  // Changing N to > 32 (the compile_time tile dim breaks the code still)
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Nit, I think this has been resolved.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 4]
```diff
@@ -24,6 +24,27 @@ LoopNestGenerator::LoopNestGenerator(
   generate(exprs);
 }
 
+namespace {
+
+// Currently, allocation of smem tennsors and indexing is
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp]: Nit: tennsors->tensors

**Review by csarofeen — APPROVED** [review]
LGTM, minor nits.

**Pull request merged**